database/skylink.go

```diff
@@ -0,0 +1,94 @@
+package database
+
+import (
+	"context"
+	"regexp"
+
+	"gitlab.com/NebulousLabs/errors"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+)
+
+var (
+	skylinkRE *regexp.Regexp = nil
+)
+
+// Skylink represents a skylink object in the DB.
+type Skylink struct {
+	ID      primitive.ObjectID `bson:"_id,omitempty" json:"-"`
+	Skylink string             `bson:"skylink" json:"skylink"`
+	Size    int64              `bson:"size" json:"size"`
+}
+
+// Skylink gets the DB object for the given skylink.
+// If it doesn't exist it creates it.
+func (db *DB) Skylink(ctx context.Context, skylink string) (*Skylink, error) {
+	skylinkHash, err := validateSkylink(skylink)
+	if err != nil {
+		return nil, ErrInvalidSkylink
+	}
+	// Provisional skylink object.
+	skylinkRec := Skylink{
+		Skylink: skylinkHash,
+	}
+	// Try to find the skylink in the database.
+	filter := bson.D{{"skylink", skylinkHash}}
+	sr := db.staticSkylinks.FindOne(ctx, filter)
+	err = sr.Decode(&skylinkRec)
+	if err == mongo.ErrNoDocuments {
+		// It's not there, insert it.
+		var ior *mongo.InsertOneResult
+		ior, err = db.staticSkylinks.InsertOne(ctx, skylinkRec)
+		if err == nil {
+			skylinkRec.ID = ior.InsertedID.(primitive.ObjectID)
+		}
+	}
+	if err != nil {
+		return nil, err
+	}
+	return &skylinkRec, nil
+}
+
+// SkylinkByID finds a skylink by its ID.
+func (db *DB) SkylinkByID(ctx context.Context, id primitive.ObjectID) (*Skylink, error) {
+	filter := bson.D{{"_id", id}}
+	sr := db.staticSkylinks.FindOne(ctx, filter)
+	var sl Skylink
+	err := sr.Decode(&sl)
+	if err != nil {
+		return nil, err
+	}
+	return &sl, nil
+}
+
+// SkylinkUpdate updates the meta data about the given skylink. If any of the
+// parameters is empty they won't be used in the update operation.
+func (db *DB) SkylinkUpdate(ctx context.Context, id primitive.ObjectID, name string, size int64) error {
+	filter := bson.M{"_id": id}
+	updates := bson.M{}
+	if name != "" {
+		updates["name"] = name
+	}
+	if size > 0 {
+		updates["size"] = size
+	}
+	_, err := db.staticSkylinks.UpdateOne(ctx, filter, bson.M{"$set": updates})
+	if err != nil {
+		return errors.AddContext(err, "failed to update")
+	}
+	return nil
+}
+
+// validateSkylink extracts the skylink hash from the given skylink that might
+// have protocol, path, etc. within it.
+func validateSkylink(skylink string) (string, error) {
+	if skylinkRE == nil {
```

Comment:
why don't we simly set the regex in the var opposed to lazy loading it like this?